src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java

```diff
@@ -761,4 +754,199 @@ public static String getSchemaPositionSHA(Long serverID, Position position) {
+
+		if (schemaID == null) {
+			throw new RuntimeException("Schema not saved yet.");
+		}
+		if (schemaChain.length() <= chainLength || schemaChain.length() < chainMaxLength) {
+			return 0;
+		}
+
+		long[] removeIds;
+		SchemaChain newSchemaChain = new SchemaChain();
+
+		connection.setAutoCommit(false);
+		try {
+			Long newBaseSchemaID;
+
+			if (chainLength == 0) {
+				// remove all base schemas
+
+				removeIds = new long[schemaChain.length()];
+				int n = 0;
+				for (long id : schemaChain) {
+					removeIds[n++] = id;
+				}
+				newBaseSchemaID = schemaID;
+				saveFullSchemaData(connection, schemaID);
+
+			} else {
+				// split chain between schemas to keep and to remove
+
+				long[] keepIds = new long[chainLength];
+				removeIds = new long[schemaChain.length() - chainLength];
+				int n = 0;
+				for (long id : schemaChain) {
+					if (n < chainLength) {
+						keepIds[n++] = id;
+					} else {
+						removeIds[n++ - chainLength] = id;
+					}
+				}
+
+				// save new base schema data
+				newBaseSchemaID = keepIds[chainLength - 1];
+
+				MysqlSavedSchema newBase = new MysqlSavedSchema(serverID, sensitivity);
+				try {
+					newBase.restoreFromSchemaID(connection, newBaseSchemaID);
+				} catch (InvalidSchemaError e) {
+					throw new RuntimeException("Couldn't restore schema " + newBaseSchemaID + ".", e);
+				}
+
+				newBase.saveFullSchemaData(connection, newBase.schemaID);
+
+				for (int i = keepIds.length - 1; i >= 0; i--) {
+					newSchemaChain.push(keepIds[i]);
+				}
+			}
+
+			// remove reference to parent of the new base
+			try (PreparedStatement stmt = connection.prepareStatement(
+				"UPDATE `schemas` SET `base_schema_id` = NULL, `deltas` = NULL WHERE `id`= ?"
+			)) {
+				stmt.setLong(1, newBaseSchemaID);
+				stmt.executeUpdate();
+			}
+
+			// clean up
+			StringBuilder inSQL = new StringBuilder();
+			for (int i = 0; i < removeIds.length; i++) {
+				if (inSQL.length() > 0) {
+					inSQL.append(", ");
+				}
+				inSQL.append("?");
+			}
+			for (String tableName : new String[]{"columns", "tables", "databases"}) {
+				try (PreparedStatement stmt = connection.prepareStatement(
+					"DELETE FROM `" + tableName + "` WHERE `schema_id` IN (" + inSQL + ")"
+				)) {
+					for (int i = 0; i < removeIds.length; i++) {
+						stmt.setLong(i + 1, removeIds[i]);
+					}
+					stmt.executeUpdate();
+				}
+			}
+
+			try (PreparedStatement stmt = connection.prepareStatement(
+				"DELETE FROM `schemas` WHERE `id` IN (" + inSQL + ")"
+			)) {
+				for (int i = 0; i < removeIds.length; i++) {
+					stmt.setLong(i + 1, removeIds[i]);
+				}
+				stmt.executeUpdate();
+			}
+
+			connection.commit();
+
+		} catch (Throwable e) {
+			connection.rollback();
+			throw e;
+
+		} finally {
+			connection.setAutoCommit(true);
+		}
+
+		schemaChain = newSchemaChain;
+		if (schemaChain.length() == 0) {
+			deltas = null;
+		}
+
+		return removeIds.length;
+	}
+
+	private static class SchemaChain implements Iterable<Long> {
```

Comment:
confused as to why we're implementing our own linked list.  shouldn't this just be `extends LinkedList<Long>`?  

Actually I'm utterly confused by the value of this, other than maybe the `id()` method....

Reply:
LinkedList's nodes are mutable, therefore creating copy of LinkedList always re-creates the whole chain. This is an optimization with immutable nodes - creating new chain reuses existing nodes, only allocates new head. Might be useless optimization. If so, I think it would be better to use array of IDs / ArrayList as they have better data locality.

Reply:
After reading it and trying to think through it a bunch, I think that `SchemaChain` as written is in a weird middle-ground; it's not doing enough that's specific to the chain, and not general enough to be just called "ImmutableLinkedList" or whatever.

I'd say, either: 
- throw it out and use generic data structures; an array is fine, *or*:
- go ahead and add *more* functionality into `SchemaChain` so that it becomes responsible for more stuff, I could easily see all the logic in `restoreDerivedSchema` living inside SchemaChain.